### PR TITLE
[coordinates.transform] Sparse sampling TICA transformer

### PR DESCRIPTION
**Pull request by florianlitzinger**
This introduces the NystroemTICA transformer that implements a sparse sampling variant of TICA based on a selection of column subsets of the relevant covariance matrices. To that end, additional arguments to coordinates.estimation.covariance, _ext.variational.estimators.moments, and _ext.variational.estimators.running_moments have been introduced that allow for the restriction of the covariance computation to the diagonal or a subset of columns, respectively. The tests have been extended to cover the new features.

### CODE REVIEW

File: pyemma/coordinates/api.py [hunk 1]
```diff
@@ -1414,9 +1415,113 @@ def vamp(data=None, lag=10, dim=None, scaling=None, right=True, ncov_max=float('
     return res
 
 
+def nystroem_tica(data, lag, max_columns,
```
Comment (franknoe) [inline pyemma/coordinates/api.py]: Please rename to ''tica_nystroem'' (makes it easier to find)

**Review by franknoe — COMMENTED** [review]
Good for me except for the minor name change indicated here. It would be good if @fabian-paul  or @fnueske could have a look too.

File: pyemma/coordinates/api.py (hunk 1 above)
Comment (florianlitzinger) [reply pyemma/coordinates/api.py]: OK, done.

**Comment by marscher** [comment]
Thanks for this major contribution! In general the code looks also good to me. Here are my 50 cents:

A lot of code has been changed, but unfortunately not tested, especially in the moments module (https://codecov.io/gh/markovmodel/PyEMMA/pull/1261/diff?src=pr&el=tree#diff-cHllbW1hL19leHQvdmFyaWF0aW9uYWwvZXN0aW1hdG9ycy9tb21lbnRzLnB5 - every line which is in a big green area has been added and if the bar on the left is red, the code has not been executed during the tests. So I doubt that this code is either needed, or you need to demonstrate the code is working by providing a test case (untested code is unfortunately most often broken code). So I guess all of the sparse
input things have not been tested.

In the test for the Nyström method you generate a pseudo sparse matrix, in which 10 out of 100 columns have a meaning. However the order of these columns is fixed, right? I would suggest to shuffle the columns with a meaning to make this test even harder.

A lot of code has been copied from the tica module (which codacy also detected). This indicates we should introduce a common base class for PCA, TICA, VAMP and now Nystroem, since they all share eg. the handling of dimensions. Lazy diagonalization is also copied all over the place. Duplicated code makes it harder to change things in the future. But I guess we need to change this, as there should be some discussion which functionality to share.

Just another minor thing, please use new-style objects for your classes (derive them from object).

File: pyemma/coordinates/api.py [hunk 2]
```diff
@@ -1414,9 +1415,113 @@ def vamp(data=None, lag=10, dim=None, scaling=None, right=True, ncov_max=float('
...
 
+def tica_nystroem(data, lag, max_columns,
+                  dim=-1, var_cutoff=0.95, epsilon=1e-6,
+                  stride=1, skip=0, reversible=True, ncov_max=float('inf'),
+                  initial_columns=None, nsel=1, neig=None):
+    r""" Sparse sampling implementation [1]_ of time-lagged independent component analysis (TICA) [2]_, [3]_, [4]_.
+
+    Parameters
+    ----------
+    data : ndarray (T, d) or list of ndarray (T_i, d) or a reader created by
```
Comment (marscher) [inline pyemma/coordinates/api.py]: the docstring says data is optional, but it isn't. All of the other estimators in coordinates api support just the construction of the object without estimation.

File: pyemma/coordinates/transform/nystroem_tica.py [hunk 3]
```diff
@@ -0,0 +1,921 @@
...
+        # check nsel
+        nsel = self._check_nsel()
+        if nsel is None:
+            return None
+        # go
+        return self._select(nsel, err)
+
+    def _select(self, nsel, err):
+        """ Override me to do selection """
+        pass
```
Comment (marscher) [inline pyemma/coordinates/transform/nystroem_tica.py]: raise NotImplementedError here, so this fails, if a deriving class forgot to overload this.

File: pyemma/coordinates/transform/nystroem_tica.py [hunk 4]
```diff
@@ -0,0 +1,921 @@
...
+
+
+class SelectionStrategyRandom(SelectionStrategy):
```
Comment (marscher) [inline pyemma/coordinates/transform/nystroem_tica.py]: test me

File: pyemma/coordinates/transform/nystroem_tica.py [hunk 5]
```diff
@@ -0,0 +1,921 @@
...
+        # do select
+        sel = []
+        while len(sel) < nsel:
+            i = np.random.choice(self._oasis_obj._n)
+            if not (i in sel or i in self._oasis_obj._columns):
+                sel.append(i)
+        return np.array(sel, dtype=int)
+
+
+class SelectionStrategyOasis(SelectionStrategy):
```
Comment (marscher) [inline pyemma/coordinates/transform/nystroem_tica.py:846]: test me

File: pyemma/_ext/variational/estimators/moments.py [hunk 6]
```diff
@@ -359,20 +359,47 @@ def _center(X, w, s, mask=None, const=None, inplace=True):
     return X, const
 
 
+def _filter_variable_indices(mask, column_selection):
```
Comment (marscher) [inline pyemma/_ext/variational/estimators/moments.py:362]: this function is never called.

File: pyemma/_ext/variational/estimators/moments.py [hunk 7]
```diff
@@ -521,14 +567,34 @@ def _M2_symmetric(Xvar, Yvar, mask_X=None, mask_Y=None, xsum=0, xconst=0, ysum=0
...
+                Xvark = Xvar
+                mask_Xk = mask_X
+                xkvarsum = xvarsum
+                xkconst = xconst
+                Yvark = Yvar
+                mask_Yk = mask_Y
+                ykvarsum = yvarsum
+                ykconst = yconst
+            else:
+                Xvark = Xvar[:, _filter_variable_indices(mask_X, column_selection)].copy()
```
Comment (marscher) [inline pyemma/_ext/variational/estimators/moments.py]: why do you take a copy here? Is the selection modified in place?

File: pyemma/coordinates/tests/test_nystroem_tica.py [hunk 8]
```diff
@@ -82,5 +90,79 @@ def test(self):
         assert np.all(np.abs(self.tica_obj.eigenvalues[:4] - t.eigenvalues[:4]) < 1e-3)
 
 
+class TestNystroemTICA_oASIS(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.X = np.ones((10000, 10))
+        cls.variable_columns = np.random.choice(10, 5, replace=False)
```
Comment (marscher) [inline pyemma/coordinates/tests/test_nystroem_tica.py:97]: Thanks! But does the test also pass if you make the choice out of np.random.choice(cls.X.shape[1], 5, replace=False)

**Comment by marscher** [comment]
Thanks for the bugfixes and additional tests. The Codacy issues are mostly about copy paste code, but I think we can restructure things after merging, since it requires some additional design discussions. The issue about the not initialized member variable is actually a false positive.

File: pyemma/coordinates/tests/test_nystroem_tica.py (hunk 8 above)
Comment (florianlitzinger) [reply pyemma/coordinates/tests/test_nystroem_tica.py:97]: How do you mean? (cls.X.shape[1] == 10)

**Comment by marscher** [comment]
Indeed, makes no sense. Can you blow up the matrix to a higher dimension 
(eg. 100 or even more) and place 10 meaningful columns at random indices?

**Pull request merged**